### ISSUE TITLE
Redirect to product page if only one result doesn't work in case of void return type in the controller.

### ISSUE THREAD

**Issue by yutv** [Smile-SA/elasticsuite#2383]
### Preconditions

Magento Version : 2.4.2 EE

ElasticSuite Version : 2.10.7

Environment : Developer

Third party modules : a lot but it doesn't matter

### Steps to reproduce

1. create a preference for the `Magento\CatalogSearch\Controller\Result\Index `
    ```php
    <?php
    /**
     * Result Index Action
     */

    declare(strict_types=1);

    namespace SomeVendor\CatalogSearch\Rewrite\Controller\Result;

    use Magento\CatalogSearch\Controller\Result\Index as MagentoIndex;

    /**
     * Result Index Action
     */
    class Index extends MagentoIndex
    {
        public function execute(): void
        {
            parent::execute();
            // some other stuff goes here
        }
    }
    ```
2. Enable redirect for one product result: 
    **Stores > Configuration > Elasticsuite > Catalog Search > Catalog Search Configuration > Redirect to product page if only one result = Yes**
3. Proceed with a search which returns one product

### Expected result

1. Redirect to a product detail page

### Actual result

1. Search result page with one product

### More details

The native `Magento\CatalogSearch\Controller\Result\Index::execute` method is marked as `void` in the dockbock so it was logical to use `void` as a function return type in the preference. As the result the following interceptor code was generated (please pay attention there is no `return` statement, which is logical since we have the `void` return type in the method)
```php
generated/code/SomeVendor/CatalogSearch/Rewrite/Controller/Result/Index/Interceptor.php
...
    public function execute() : void
    {
        $pluginInfo = $this->pluginList->getNext($this->subjectType, 'execute');
        $pluginInfo ? $this->___callPlugins('execute', func_get_args(), $pluginInfo) : parent::execute();
    }
```
instead of
```php
    public function execute()
    {
        $pluginInfo = $this->pluginList->getNext($this->subjectType, 'execute');
        return $pluginInfo ? $this->___callPlugins('execute', func_get_args(), $pluginInfo) : parent::execute();
    }
```

Eventually the `Result\Redirect` returned in the `vendor/smile/elasticsuite/src/module-elasticsuite-catalog/Plugin/CatalogSearch/ResultPlugin.php:139` was lost.

So the elasticsuite module just doesn't work properly in case controller action is not designed to return a `ResultInterface`.

**Comment by romainruaud** [comment]
I'm not sure what we could change here to be honest.

You should just not add a return type to the `execute()` function of your preference.

How could we improve this ?

**Comment by yutv** [comment]
The change which can be done is to cover both cases. The following patch resolves the issue.

```patch
diff '--exclude=.git' -ruN a/src/module-elasticsuite-catalog/Plugin/CatalogSearch/ResultPlugin.php b/src/module-elasticsuite-catalog/Plugin/CatalogSearch/ResultPlugin.php
--- a/src/module-elasticsuite-catalog/Plugin/CatalogSearch/ResultPlugin.php	2021-12-09 20:35:17.187735596 +0200
+++ b/src/module-elasticsuite-catalog/Plugin/CatalogSearch/ResultPlugin.php	2021-12-09 20:38:33.925514488 +0200
@@ -17,6 +17,7 @@
 use Magento\Catalog\Model\Layer\Resolver;
 use Magento\CatalogSearch\Helper\Data;
 use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Framework\App\ResponseInterface;
 use Magento\Framework\Controller\ResultFactory;
 use Magento\Framework\Message\ManagerInterface;
 use Magento\Framework\Event\ManagerInterface as EventManagerInterface;
@@ -84,7 +85,8 @@
         Data $catalogSearchHelper,
         ManagerInterface $messageManager,
         ResultFactory $resultFactory,
-        EventManagerInterface $eventManager
+        EventManagerInterface $eventManager,
+        ResponseInterface $response
     ) {
         $this->layerResolver  = $layerResolver;
         $this->scopeConfig    = $scopeConfig;
@@ -92,6 +94,7 @@
         $this->helper         = $catalogSearchHelper;
         $this->resultFactory  = $resultFactory;
         $this->eventManager   = $eventManager;
+        $this->reponse        = $response;
     }
 
     /**
@@ -123,6 +126,7 @@
                         $this->addRedirectMessage($product);
                         $result = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
                         $result->setUrl($product->getProductUrl());
+                        $this->reponse->setRedirect($product->getProductUrl());
 
                         $this->eventManager->dispatch(
                             'smile_elasticsuite_redirect_if_one_result',

```

**Comment by romainruaud** [comment]
can you propose a PR containing this fix ?

Regards

**Comment by romainruaud** [comment]
Fixed by PR #2395
